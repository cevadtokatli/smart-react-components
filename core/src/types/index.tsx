export type ContentElement = string | JSXChildren

export type JSXChildren = JSX.Element | JSX.Element[]

export interface JSXElementProps {
  [key: string]: any
}

export interface NestedObject<T> {
  [key: string]: T | NestedObject<T>
}

export type PaletteProp = 'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info' | 'light' | 'dark' | (string & {})

export type Partial<T> = {
  [key in keyof T]?: T[key] extends object ? Partial<T[key]> : T[key]
}

export type ResponsiveProp<K extends string, T> = {
  [key in K as K]: T
} & {
  [key in K as `${K}Sm` | `${K}Md` | `${K}Lg` | `${K}Xl`]?: T
}

export type ResponsiveProps<T> =
{ [key in keyof T as key]?: T[key] }
& { [key in keyof T as key extends string ? `${key}Sm` : never]?: T[key] }
& { [key in keyof T as key extends string ? `${key}Md` : never]?: T[key] }
& { [key in keyof T as key extends string ? `${key}Lg` : never]?: T[key] }
& { [key in keyof T as key extends string ? `${key}Xl` : never]?: T[key] }

export interface Size<T> {
  small: T
  medium: T
  large: T
  [key: string]: T
}

export type SizeProp = 'small' | 'medium' | 'large' | (string & {})

export type Value = string | number
