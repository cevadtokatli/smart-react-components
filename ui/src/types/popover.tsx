import { Position } from './'

export interface PopoverCalculationResult {
  arrowStyle: string
  isArrowInHeader: boolean
  overflow: number
  pos: Position
  style: string
}
