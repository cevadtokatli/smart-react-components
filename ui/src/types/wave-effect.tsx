export interface EffectItem {
  element: JSX.Element
  endTime: number
}
