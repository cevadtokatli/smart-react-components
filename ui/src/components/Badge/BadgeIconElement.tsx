import Div from '@smart-react-components/core/Element/Div'
import { StyledProps } from '@smart-react-components/core/styled-props'
import { PaletteProp, ShapeProp } from '@smart-react-components/core/types'
import styled from 'styled-components'
import { OrderPosition } from '../../types'
import { getReverseOrderPosition } from '../../util/props'

interface Props extends StyledProps {
  iconPosition: OrderPosition
  isClickable: boolean
  isOutline: boolean
  isSoft: boolean
  palette: PaletteProp
  shape: ShapeProp
}

export default styled(Div)<Props>(({ theme, iconPosition, isClickable, isOutline, isSoft, palette, shape }: Props) => `
  align-items: center;
  align-self: stretch;
  border: solid 1px;
  border-${OrderPosition[getReverseOrderPosition(iconPosition)].toLowerCase()}: 0;
  border-color: ${!isSoft ? theme.$.palette[palette].dynamicer : theme.$.palette[palette].softDarker};  
  border-radius: ${theme.$.radius.badge[shape]};
  border-bottom-${OrderPosition[getReverseOrderPosition(iconPosition)].toLowerCase()}-radius: 0;
  border-top-${OrderPosition[getReverseOrderPosition(iconPosition)].toLowerCase()}-radius: 0;
  display: inline-flex;
  flex: 0 0 auto;
  justify-content: center;

  ${!isOutline
    ? `  
      ${!isSoft
        ? `
          background: ${theme.$.palette[palette].dynamicer};
          color: ${theme.$.palette[palette].font};
          fill: ${theme.$.palette[palette].font};
        `
        : `
          background: ${theme.$.palette[palette].softDarker};
          color: ${theme.$.palette[palette].softFont};
          fill: ${theme.$.palette[palette].softFont};       
        `
      }

      ${isClickable
        ? `
          cursor: ${theme.$.cursor.badgeIconClickable};

          &:active {
            ${!isSoft
              ? `
                background: ${theme.$.palette[palette].dynamicest};
                border-color: ${theme.$.palette[palette].dynamicest};  
              `
              : `
                background: ${theme.$.palette[palette].softDarkest};
                border-color: ${theme.$.palette[palette].softDarkest};
              `
            }
          }
        `
        : ''
      }
    `
    : `
     ${!isSoft
        ? `
          color: ${theme.$.palette[palette].main};
          fill: ${theme.$.palette[palette].main};
        `
        : `
          color: ${theme.$.palette[palette].soft};
          fill: ${theme.$.palette[palette].soft};       
        `
      }

      ${isClickable
        ? `
          &:active {
            ${!isSoft
              ? `
                background: ${theme.$.palette[palette].dynamicer};
                color: ${theme.$.palette[palette].font};
                fill: ${theme.$.palette[palette].font};
              `
              : `
                background: ${theme.$.palette[palette].softDarker};
                color: ${theme.$.palette[palette].softFont};
                fill: ${theme.$.palette[palette].softFont};
              `
            }
          }
        `
        : ''
      }
    `
  }
`)
