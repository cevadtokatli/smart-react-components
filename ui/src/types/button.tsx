import { IntrinsicStyledCoreProps } from '@smart-react-components/core/element-props/intrinsic-styled-core-props'
import { ContentElement, JSXElementProps, PaletteProp, Partial, ResponsiveProp, ShapeProp, SizeProp } from '@smart-react-components/core/types'

export interface ButtonGenericProps<T extends ContentElement> extends
  IntrinsicStyledCoreProps,
  Partial<ResponsiveProp<'size', SizeProp>> {
  children: T | T[]
  elementProps?: JSXElementProps
  hasHover?: boolean
  hasSpace?: boolean
  hasWaveEffect?: boolean
  isActive?: boolean
  isBlock?: boolean
  isDisabled?: boolean
  isFixedSize?: boolean
  isLink?: boolean
  isLoading?: boolean
  isSoft?: boolean
  isOutline?: boolean
  leftAddon?: JSX.Element
  loading?: JSX.Element
  palette?: PaletteProp
  rightAddon?: JSX.Element
  shape?: ShapeProp
  waveEffectPalette?: PaletteProp
}

export interface ButtonListGenericProps extends
  IntrinsicStyledCoreProps,
  Partial<ResponsiveProp<'size', SizeProp>> {
  children: JSX.Element[]
  elementProps?: JSXElementProps
  hasHover?: boolean
  hasWaveEffect?: boolean
  isBlock?: boolean
  isFixedSize?: boolean
  isLink?: boolean
  isSoft?: boolean
  isOutline?: boolean
  palette?: PaletteProp
  shape?: ShapeProp
  waveEffectPalette?: PaletteProp
}

export enum ButtonType {
  BUTTON = 'button',
  RESET = 'reset',
  SUBMIT = 'submit',
}
