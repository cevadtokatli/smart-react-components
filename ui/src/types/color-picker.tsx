import { Color } from '@smart-react-components/core/types'
import { MutableRefObject } from 'react'

export enum ColorPickerFormat {
  HEX,
  RGB,
  HSL,
}

export interface PickerProps {
  color: MutableRefObject<Color>
  onDrag: () => void
  onEnd: (onMove: (e: Event) => void, onEnd: () => void) => void
  onRef?: (v: PickerRef) => void
  onStart: (onMove: (e: Event) => void, onEnd: () => void) => void
}

export interface PickerRef {
  onChange: () => void
  onExternalDrag?: () => void
}
