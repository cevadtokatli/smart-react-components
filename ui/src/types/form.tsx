import { IntrinsicStyledProps } from '@smart-react-components/core/element-props/intrinsic-styled-props'
import { ContentElement, JSXElementProps, PaletteProp, Partial, ResponsiveProp, SetState, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import { FormValue } from '../types'

export interface GenericProps extends
  Partial<ResponsiveProp<'size', SizeProp>>,
  IntrinsicStyledProps {
  active: FormValue | FormValue[]
  children: JSX.Element | JSX.Element[]
  elementProps?: JSXElementProps
  hasBorder?: boolean
  hasHover?: boolean
  hasWaveEffect?: boolean
  isBlock?: boolean
  isDisabled?: boolean
  isRequired?: boolean
  isOutline?: boolean
  isSoft?: boolean
  label?: ContentElement
  leftAddon?: ContentElement
  leftIcon?: ContentElement
  palette?: PaletteProp
  rightAddon?: ContentElement
  rightIcon?: ContentElement
  shape?: ShapeProp
  setActive: SetState<FormValue | FormValue[]>
  waveEffectPalette?: PaletteProp
}
