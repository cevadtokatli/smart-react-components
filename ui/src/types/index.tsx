import { Value } from '@smart-react-components/core/types'

export type FormValue = Value | boolean

export enum InputType {
  EMAIL = 'email',
  PASSWORD = 'password',
  SEARCH = 'search',
  TEL = 'tel',
  TEXT = 'text',
  URL = 'url',
}

export enum OrderPosition {
  LEFT = 1,
  RIGHT = 2,
}

export enum Position {
  LEFT = 1,
  RIGHT = 2,
  TOP = 4,
  BOTTOM = 8,
}

export enum TriggerInteraction {
  NONE = 0,
  CLICK = 1,
  HOVER = 2,
  RIGHT_CLICK = 4,
}
