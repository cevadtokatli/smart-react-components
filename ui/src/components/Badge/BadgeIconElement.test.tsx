import { wrapTheme } from '@smart-react-components/core/test'
import { render } from '@testing-library/react'
import React from 'react'
import { OrderPosition } from '../../types'
import BadgeIconElement from './BadgeIconElement'

describe('<BadgeIconElement />', () => {
  let props

  beforeAll(() => {
    props = {
      iconPosition: OrderPosition.LEFT,
      isClickable: false,
      isOutline: false,
      isSoft: false,
      palette: 'primary',
      shape: 'rectangle',
    }
  })

  it('should render component', () => {
    const { asFragment } = render(wrapTheme(<BadgeIconElement {...props} />))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render component with soft colors', () => {
    const { asFragment } = render(wrapTheme(<BadgeIconElement {...props} isSoft />))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render component outline', () => {
    const { asFragment } = render(wrapTheme(<BadgeIconElement {...props} isOutline />))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render component outline with soft colors', () => {
    const { asFragment } = render(wrapTheme(<BadgeIconElement {...props} isOutline isSoft />))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should be clickable', () => {
    const { asFragment } = render(wrapTheme(<BadgeIconElement {...props} isClickable />))
    expect(asFragment()).toMatchSnapshot()
  })
})
