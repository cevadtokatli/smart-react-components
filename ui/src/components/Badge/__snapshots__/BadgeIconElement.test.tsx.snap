// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BadgeIconElement /> should be clickable 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  border: solid 1px;
  border-right: 0;
  border-color: #0556B2;
  border-radius: .15rem;
  border-bottom-right-radius: 0;
  border-top-right-radius: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #0556B2;
  color: #fff;
  fill: #fff;
  cursor: pointer;
}

.c0:active {
  background: #0552A9;
  border-color: #0552A9;
}

<div
    class=" c0 src-badge-icon"
    shape="rectangle"
  />
</DocumentFragment>
`;

exports[`<BadgeIconElement /> should render component 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  border: solid 1px;
  border-right: 0;
  border-color: #0556B2;
  border-radius: .15rem;
  border-bottom-right-radius: 0;
  border-top-right-radius: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #0556B2;
  color: #fff;
  fill: #fff;
}

<div
    class=" c0 src-badge-icon"
    shape="rectangle"
  />
</DocumentFragment>
`;

exports[`<BadgeIconElement /> should render component outline 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  border: solid 1px;
  border-right: 0;
  border-color: #0556B2;
  border-radius: .15rem;
  border-bottom-right-radius: 0;
  border-top-right-radius: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #0665d0;
  fill: #0665d0;
}

<div
    class=" c0 src-badge-icon"
    shape="rectangle"
  />
</DocumentFragment>
`;

exports[`<BadgeIconElement /> should render component outline with soft colors 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  border: solid 1px;
  border-right: 0;
  border-color: #a7c9ee;
  border-radius: .15rem;
  border-bottom-right-radius: 0;
  border-top-right-radius: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #cde0f5;
  fill: #cde0f5;
}

<div
    class=" c0 src-badge-icon"
    shape="rectangle"
  />
</DocumentFragment>
`;

exports[`<BadgeIconElement /> should render component with soft colors 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  border: solid 1px;
  border-right: 0;
  border-color: #a7c9ee;
  border-radius: .15rem;
  border-bottom-right-radius: 0;
  border-top-right-radius: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #a7c9ee;
  color: #033c7c;
  fill: #033c7c;
}

<div
    class=" c0 src-badge-icon"
    shape="rectangle"
  />
</DocumentFragment>
`;
